src/verify.rs

```diff
@@ -0,0 +1,615 @@
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum SchemeID {
+    PedersenBlsChained,
+    PedersenBlsUnchained,
+    UnchainedOnG1,
+    UnchainedOnG1RFC9380,
+}
+
+impl<'de> Deserialize<'de> for SchemeID {
+    fn deserialize<D>(deserializer: D) -> Result<SchemeID, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = Deserialize::deserialize(deserializer)?;
+        match s {
+            "pedersen-bls-chained" => Ok(SchemeID::PedersenBlsChained),
+            "pedersen-bls-unchained" => Ok(SchemeID::PedersenBlsUnchained),
+            "bls-unchained-on-g1" => Ok(SchemeID::UnchainedOnG1),
+            "bls-unchained-g1-rfc9380" => Ok(SchemeID::UnchainedOnG1RFC9380),
+            _ => Err(serde::de::Error::unknown_variant(
+                s,
+                &[
+                    "pedersen-bls-chained",
+                    "pedersen-bls-unchained",
+                    "bls-unchained-on-g1",
+                    "bls-unchained-g1-rfc9380",
+                ],
+            )),
+        }
+    }
+}
+
+#[derive(Error, Debug, PartialEq)]
+pub enum VerificationError {
+    #[error("unchained schemes cannot contain a `previous_signature`")]
+    UnchainedHasPreviousSignature,
+    #[error("chained beacons must have a `previous_signature`")]
+    ChainedBeaconNeedsPreviousSignature,
+    #[error("invalid signature length")]
+    InvalidSignatureLength,
+    #[error("invalid public key")]
+    InvalidPublicKey,
+    #[error("message can't be empty")]
+    EmptyMessage,
+    #[error("signature verification failed")]
+    SignatureFailedVerification,
+    #[error("the randomness for the beacon did not match the signature")]
+    InvalidRandomness,
+}
+
+pub fn verify_beacon(
+    scheme_id: &SchemeID,
+    public_key: &[u8],
+    beacon: &Beacon,
+) -> Result<(), VerificationError> {
+    if Sha256::digest(&beacon.signature).to_vec() != beacon.randomness {
+        return Err(VerificationError::InvalidRandomness);
+    }
+    match scheme_id {
+        SchemeID::PedersenBlsChained => verify_on_g2(
+            public_key,
+            &chained_beacon_message(beacon)?,
+            &beacon.signature,
+            DST_G2,
+        ),
+        SchemeID::PedersenBlsUnchained => verify_on_g2(
+            public_key,
+            &unchained_beacon_message(beacon)?,
+            &beacon.signature,
+            DST_G2,
+        ),
+        SchemeID::UnchainedOnG1 => verify_on_g1(
+            public_key,
+            &unchained_beacon_message(beacon)?,
+            &beacon.signature,
+            DST_G2,
+        ),
+        SchemeID::UnchainedOnG1RFC9380 => verify_on_g1(
+            public_key,
+            &unchained_beacon_message(beacon)?,
+            &beacon.signature,
+            DST_G1,
+        ),
+    }
+}
+
+fn unchained_beacon_message(beacon: &Beacon) -> Result<Vec<u8>, VerificationError> {
+    if !beacon.previous_signature.is_empty() {
+        return Err(VerificationError::UnchainedHasPreviousSignature);
+    }
```

Comment:
Does this really warrants a verification error? If a user needs the previous one and decided to bundle it in the appropriate field of a Beacon even tho they are using the unchained mode, this might not be wished

Reply:
Hmm I guess it's not an error, though it's suspicous